src/signOpCert.c

```diff
@@ -0,0 +1,169 @@
+
+	{
+		TRACE_BUFFER(wireDataBuffer, wireDataSize);
+		read_view_t view = make_read_view(wireDataBuffer, wireDataBuffer + wireDataSize);
+
+		STATIC_ASSERT(SIZEOF(ctx->kesPublicKey) == KES_PUBLIC_KEY_LENGTH, "wrong KES public key size");
+		VALIDATE(view_remainingSize(&view) >= KES_PUBLIC_KEY_LENGTH, ERR_INVALID_DATA);
+		view_memmove(ctx->kesPublicKey, view.ptr, KES_PUBLIC_KEY_LENGTH);
+
+		VALIDATE(view_remainingSize(&view) >= 8, ERR_INVALID_DATA);
+		ctx->kesPeriod = parse_u8be(&view);
+
+		VALIDATE(view_remainingSize(&view) >= 8, ERR_INVALID_DATA);
+		ctx->issueCounter = parse_u8be(&view);
+
+		view_skipBytes(&view, bip44_parseFromWire(&ctx->poolColdKeyPathSpec, VIEW_REMAINING_TO_TUPLE_BUF_SIZE(&view)));
+
+		ASSERT(view_remainingSize(&view) == 0);
+	}
+
+	// Check security policy
+	security_policy_t policy = policyForSignOpCert(&ctx->poolColdKeyPathSpec);
+	ENSURE_NOT_DENIED(policy);
+
+	{
+		uint8_t opCertBodyBuffer[OP_CERT_BODY_LENGTH];
+		write_view_t opCertBodyBufferView = make_write_view(opCertBodyBuffer, opCertBodyBuffer + OP_CERT_BODY_LENGTH);
+
+		view_appendData(&opCertBodyBufferView, (const uint8_t*) &ctx->kesPublicKey, SIZEOF(ctx->kesPublicKey));
+		view_appendData(&opCertBodyBufferView, (uint8_t*) &ctx->issueCounter, sizeof(ctx->issueCounter));
+		view_appendData(&opCertBodyBufferView, (uint8_t*) &ctx->kesPeriod, sizeof(ctx->kesPeriod));
+
+		ASSERT(view_processedSize(&opCertBodyBufferView) == OP_CERT_BODY_LENGTH);
+
+		getOpCertSignature(
+		        & ctx->poolColdKeyPathSpec,
+		        opCertBodyBuffer,
+		        OP_CERT_BODY_LENGTH,
+		        ctx->signature,
+		        SIZEOF(ctx->signature)
+		);
+	}
+	ctx->responseReadyMagic = RESPONSE_READY_MAGIC;
+
+	switch (policy) {
+#	define  CASE(policy, step) case policy: {ctx->ui_step = step; break;}
+		CASE(POLICY_PROMPT_WARN_UNUSUAL,    UI_STEP_WARNING);
+		CASE(POLICY_PROMPT_BEFORE_RESPONSE, UI_STEP_CONFIRM_START);
+		CASE(POLICY_ALLOW_WITHOUT_PROMPT,   UI_STEP_RESPOND);
+#	undef   CASE
+	default:
+		ASSERT(false);
+	}
+	signOpCert_ui_runStep();
+}
+
+static void signOpCert_ui_runStep()
+{
+	ui_callback_fn_t* this_fn = signOpCert_ui_runStep;
+
+	UI_STEP_BEGIN(ctx->ui_step);
+
+	UI_STEP(UI_STEP_WARNING) {
+		ui_displayPaginatedText(
+		        "Unusual request",
+		        "Proceed with care",
+		        this_fn
+		);
+	}
+	UI_STEP(UI_STEP_CONFIRM_START) {
+		ui_displayPrompt(
+		        "Start new",
+		        "operational certificate?",
+		        this_fn,
+		        respond_with_user_reject
+		);
+	}
+	UI_STEP(UI_STEP_DISPLAY_POOL_COLD_KEY_PATH) {
+		ui_displayPathScreen("Pool cold key path", &ctx->poolColdKeyPathSpec, this_fn);
+	}
+	UI_STEP(UI_STEP_DISPLAY_KES_PUBLIC_KEY) {
+		ui_displayHexBufferScreen(
+		        "KES public key",
+		        ctx->kesPublicKey, SIZEOF(ctx->kesPublicKey),
+		        this_fn
+		);
+	}
+	UI_STEP(UI_STEP_DISPLAY_KES_PERIOD) {
+		char kesPeriodString[50];
+		snprintf(kesPeriodString, SIZEOF(kesPeriodString), "%llu", ctx->kesPeriod);
```

Comment:
support for printf formats is limited on ledger; I think we only have %u and %d for uint32_t and int, respectively
so this won't work for integers not fitting into 32 bits and we need some more or less ugly hack (see e.g. what we do with tx ttl)

Reply:
right, sorry I missed that while testing, added `str_formatUint64()` helper for that